### PR TITLE
Make artifact-cookbook more idempotent

### PR DESCRIPTION
**Pull request by KAllan357**
- Use a checksum when we talk to Nexus as well.
- Do a better check to see if someone has mucked around with an already deployed artifact.

### CODE REVIEW

File: providers/deploy.rb [hunk 1]
```diff
@@ -157,7 +159,25 @@ def delete_release_path_for(version)
...
+    current_manifest = create_manifest(current_path)
+    differences = original_manifest.find do |key, value|
+      !current_manifest.has_key?(key) || value != current_manifest[key]
+    end
+    if differences
+      Chef::Log.info "Differences found between the saved manifest at directory: #{previous_release_path} and manifest created from files at: #{current_path}. Redepoying."
+      return false
+    else
+      return true
+    end
```
Comment (KAllan357) [inline providers/deploy.rb:181]: I'm a bit worried about this sorta goofy/confusing code. Not sure how to make it more clear.

If `current_manifest` either doesn't have they keys in `original_manifest` or the same value for a key, they are found. When found returns with a value, that means there are differences. When found is `nil`, that means there were no differences so return true.

File: providers/deploy.rb [hunk 2]
```diff
@@ -157,7 +159,25 @@ def delete_release_path_for(version)
   end
 
   def deployed?
-    ::File.exists?(completion_token_path)
+    require 'yaml'
+    begin
+      Chef::Log.info "Loading manifest.yaml file from directory: #{previous_release_path}"
+      original_manifest = YAML.load_file(::File.join(previous_release_path, "manifest.yaml"))
+    rescue Errno::ENOENT, TypeError
```
Comment (reset) [inline providers/deploy.rb]: TypeError isn't a file not found

File: providers/deploy.rb [hunk 3]
```diff
@@ -157,7 +159,25 @@ def delete_release_path_for(version)
   end
 
   def deployed?
-    ::File.exists?(completion_token_path)
+    require 'yaml'
+    begin
```
Comment (reset) [inline providers/deploy.rb]: I would check to see if the file exists and then raise if it doesn't instead of rescuing and raising

**Comment by reset** [comment]
provided this all works, this looks good

File: providers/deploy.rb [hunk 4]
```diff
@@ -157,7 +161,26 @@ def delete_release_path_for(version)
...
+      raise Chef::Artifact::ManifestFileNotFound if previous_release_path.nil? || ::File.exists?(::File.join(previous_release_path, "manifest.yaml"))
```
Comment (reset) [inline providers/deploy.rb]: no, you don't want to use exceptions for control flow. Just take this out of the begin; rescue; end block

``` ruby
if previous_release_path.nil? || ::File.exists?(::File.join(previous_release_path, "manifest.yaml"))
  Chef::Log.warn "whatever message here"
  return false
end

Chef::Log.info "Loading manifest.yaml file from directory: #{previous_release_path}"
original_manifest = YAML.load_file(::File.join(previous_release_path, "manifest.yaml"))
return true
```

**Comment by reset** [comment]
if the addition of before_extract solves your use-case, :+1:

**Pull request merged**